Automattic/wp-calypso
# Checkout: Show taxes (Placeholder Edition)

This PR starts the process of adding checkout taxes.

We don't have real data available yet, and several important questions are still open, so the functionality is hidden behind the `show-tax` config flag. We're not ready to enable these even for the development config, so for now enable these using the flags query parameter, like `?flags=show-tax`


The transaction store is one of the remaining Flux stores ( see https://github.com/Automattic/wp-calypso/issues/24217 ). I'd love to find time to come back and reduxify it, but for I'm using `lib/redux-bridge` to forward the `FLUX_TRANSACTION_PAYMENT_SET` and  `TRANSACTION_NEW_CREDIT_CARD_DETAILS_SET` actions from Flux to Redux. The Payment country field was already wired up to a redux action, but this doesn't seem to cause any problems.

We're using a `httpData` request for the api call, and this effectively memoizes the request. For testing purposes, I've set a freshness of 2 days (to pluck a number out of the air). This means that you should only be able to make one successful request with a specific countryCode + postalCode combination. If you switch back to a previously used combination it won't fire off any extra requests, and once a value comes back you still won't get more requests (inside that 2 day period), you should just get the previously fetched value instantly.

Curiously, the `httpData` behaviour also seems to make it unnecessary to store the value in the redux state. I'm not quite sure what I think about that, but it's working, so...

We're calling out to the `/sites/example.wordpress.com/hello` endpoint until the tax-rate endpoint is ready in order to provide a http request to make the data-layer mechanisms run smoothly, and feeding in a dummy result if/when it returns successfully. The dummy result is the last three digits of the postcode after the decimal point (e.g. 12345 gives a 0.345 tax rate - 34.5%. An example real tax rate is 0.085).

**Testing Instructions**:

We're looking specifically for the `calypso_tax_rate_request` tracks event, along with a network request (to the placeholder address), but there are a bunch of things we want to check:

- Check that we do generate a tax event/network request (fetch) on 5 digit postcodes in the US
- Check that we never fetch with more or less than 5 digits, nor with anything other than numbers in there (it's fine if this includes valid postcodes for other countries, we're only concerned with the US for now)
- Check that we never fetch outside the US
- Check that we get both the country code and the postcode from stored credit cards.
- Check that we respond to the "New Credit/Debit Card" dialog as well as stored cards
- check that the `show-tax` config flag is required to see the tax rates, but not to trigger the http requests ( add url params: `?flags=show-tax`).

The other major concern is that we don't break anything else.

Two tricks to make this easier to verify, as shown in this gif:
1) `localStorage.setItem( 'debug', localStorage.debug + ',calypso:analytics:tracks' )`, and then optionally filter the log on `calypso_tax_rate_request`
2) pin `state.ui.payment` in redux tools so you can watch the values.
3) Change the `freshness` value in `requestTaxRate()` in `client/state/data-getters/index.js` to `60  * 1000` ( 1 minute) - long enough to see it work, but short enough to re-check against your stored credit cards if you want

Here, I've got 4 test credit-card pre-populated, 2 with different US postcodes, 1 from `AU` with a postcode that wouldn't match anyway, and one from `CA` with a 5 digit postcode:

![checkout httpdata](https://user-images.githubusercontent.com/5952255/47715274-f9d6c680-dc8a-11e8-86a3-305da2127ab6.gif)
